lib/Sema/DerivedConformanceCodable.cpp

```diff
@@ -31,147 +31,154 @@ using namespace DerivedConformance;
 ///
 /// \param proto The {En,De}codable protocol to validate all the keys conform
 /// to.
-static bool validateCodingKeysEnum(TypeChecker &tc, EnumDecl *codingKeysDecl,
-                                   NominalTypeDecl *type, ProtocolDecl *proto) {
+static bool
+validateCodingKeysEnum(TypeChecker &tc, EnumDecl *codingKeysDecl,
+                       NominalTypeDecl *target, ProtocolDecl *proto) {
   // Look through all var decls in the given type.
   // * Filter out lazy/computed vars (currently already done by
   //   getStoredProperties).
   // * Filter out ones which are present in the given decl (by name).
   //
   // If any of the entries in the CodingKeys decl are not present in the type
   // by name, then this decl doesn't match.
-  // If there are any vars left in the type, then this decl doesn't match.
-  //
-  // NOTE: If we change the behavior to ignore vars with default values, then we
-  //       can further filter out the type names to remove those which
-  //       correspond to vars with default values.
-  llvm::SmallDenseSet<Identifier, 8> names;
-
-  auto storedProperties = type->getStoredProperties(/*skipInaccessible=*/true);
-  auto validVarCallback = [&names](VarDecl *varDecl) {
-    names.insert(varDecl->getName());
-  };
-
-  if (!validateVarsConformToProtocol(tc, type->getDeclContext(),
-                                     storedProperties, proto, validVarCallback))
-    return false;
+  // If there are any vars left in the type which don't have a default value
+  // (for Decodable), then this decl doesn't match.
+
+  // Here we'll hold on to properties by name -- when we've validated a property
+  // against its CodingKey entry, it will get removed.
+  llvm::SmallDenseMap<Identifier, VarDecl *, 8> properties;
+  for (auto *varDecl : target->getStoredProperties(/*skipInaccessible=*/true)) {
+    properties[varDecl->getName()] = varDecl;
+  }
 
+  bool propertiesAreValid = true;
   for (auto elt : codingKeysDecl->getAllElements()) {
-    auto it = names.find(elt->getName());
-    if (it == names.end()) {
-      // TODO: Produce diagnostic here complaining that the CodingKeys enum
-      //       contains a case which does not correspond to a var.
+    auto it = properties.find(elt->getName());
+    if (it == properties.end()) {
+      tc.diagnose(elt->getLoc(), diag::codable_extraneous_codingkey_case_here,
+                  elt->getName());
       // TODO: Investigate typo-correction here; perhaps the case name was
       //       misspelled and we can provide a fix-it.
-      return false;
+      propertiesAreValid = false;
+      continue;
+    }
+
+    // We have a property to map to. Ensure it's {En,De}codable.
+    bool conforms = varConformsToProtocol(tc, target->getDeclContext(),
+                                          it->second, proto);
+    if (!conforms) {
+      tc.diagnose(it->second->getLoc(),
+                  diag::codable_non_conforming_property_here,
```

Comment:
Teeechnically the property might have had some other error if it failed to validate, which means this diagnostic will be spurious. Most parts of the compiler don't bother to handle that kind of thing, though.